src/commonMain/kotlin/komem/litmus/runners/WorkerTestRunner.kt

```diff
@@ -1,69 +1,91 @@
+        data class WorkerContext(
+            val states: List<S>,
+            val threadOutcomes: MutableList<Any?>?,
+            val threadFunction: S.() -> Any?,
+            val syncPeriod: Int,
+            val barrier: Barrier,
         )
-        val futures = threadFunctions.mapIndexed { i, threadFun ->
-            val worker = Worker.start()
 
-            val affinityMap = params.affinityMap
-            if (affinityMap != null) {
+        val states = List(params.batchSize) { test.stateProducer() }
+        val barrier = params.barrierProducer(test.threadCount)
+        val outcomeFinalizer = test.outcomeFinalizer
+
+        val workers = List(test.threadCount) { Worker.start() }
+        val outcomesLists = workers.map {
+            if (outcomeFinalizer == null) MutableList<Any?>(params.batchSize) {} else null
+        }
+
+        workers.mapIndexed { threadIndex, worker ->
+            params.affinityMap?.let { affinityMap ->
                 getAffinityManager()?.run {
-                    val cpuSet = affinityMap.allowedCores(i)
+                    val cpuSet = affinityMap.allowedCores(threadIndex)
                     setAffinity(worker, cpuSet)
                     require(getAffinity(worker) == cpuSet) { "affinity setting failed" }
                 }
             }
+            val workerContext = WorkerContext(
+                states,
+                outcomesLists[threadIndex],
+                test.threadFunctions[threadIndex],
+                params.syncPeriod,
+                barrier,
+            )
             worker.execute(
                 TransferMode.SAFE /* ignored */,
-                { threadFun to workerContext }
-            ) { (threadFun, workerContext) ->
-                workerContext.apply {
-                    var cnt = 0
-                    for (test in tests) {
-                        if (cnt == this.syncPeriod) {
-                            cnt = 0
-                            barrier.wait()
-                        }
-                        threadFun(test)
-                        cnt++
+                { workerContext }
+            ) { (states, threadOutcomes, threadFunction, syncPeriod, barrier) ->
+                if (threadOutcomes != null) {
+                    for (stateIdx in states.indices) {
+                        if (stateIdx % syncPeriod == 0) barrier.wait()
+                        threadOutcomes[stateIdx] = states[stateIdx].threadFunction()
                     }
+                } else {
+                    for (i in states.indices) {
+                        if (i % syncPeriod == 0) barrier.wait()
+                        states[i].threadFunction()
+                    }
+                }
+                threadOutcomes
+            }
+            worker.requestTermination()
+        }.forEach { it.result } // await all workers
+
+        // TODO: this complex logic is defined in plain text (!) in LitmusTestDsl.kt (!)
```

Comment:
I think there we should choose just one API for thread results and support only it (there are no reasons to support two APIs). 

There are several problems with the API where outcome is the result returned from `thread`:

1. It consumes additional memory to store an array with results. 
2. Returning value from function generates additional instructions (e.g. stack pop, etc), it is not just plain load. 
3. Storing result as `Any` leads to boxing. 

I think we can borrow the idea from JCStress and pass the result object as an additional argument to `thread` (see [example](https://github.com/openjdk/jcstress/blob/ce8a2c7747e0232fac66783c89ea4a58a3819e0f/jcstress-samples/src/main/java/org/openjdk/jcstress/samples/jmm/basic/BasicJMM_06_Causality.java#L79)).

Reply:
With this idea in mind, the litmus syntax could look like this:

```
litmusTest(::Data, ::IntRegisters) {
        thread { r ->
            x = 1
            r.r1 = y
        }
        thread { r -> 
            y = 1
            r.r2 = x
        }
        outcome {
            thread[0].r1 to thread[1].r2
        }
        specification {
            accepted = setOf(
                (1 to 1), (1 to 0), (0 to 1)
            )
            interesting = setOf(
                (0 to 0)
            )
        }
 }
 
 // alt. names: IntLocals, LitmusResults, etc
 // we can pre-define several classes, e.g. LitmusIntResults, LitmusLongResults, etc
class IntRegisters {
    var r1: Int = 0
    var r2: Int = 0
}
 ```

Reply:
Above, `outcome` lambda should take as receiver some object storing an array of result objects for each thread.

Reply:
The idea behind multiple ways to store the outcome is plain convenience: in most tests threads simply write one value as their part of the outcome, or just one thread writes it. It's not really handy to always write a separate section just to mark one or two variables as an outcome.
1. Memory to store the results is consumed in any case, and it's not a bottleneck either: there is no need to allocate a lot of states at once (i.e. `batchSize`<10M is a reasonable assumption because the test can be repeated)
2. But this still is a function and it still will do stack pops and stuff? As far as I remember from my OS course (which was obviously just the tip of the iceberg), returning a value is as simple as writing it to a specific register?
3. Boxing here shouldn't be a problem in terms of possible results, because the box is not shared

Creating a lot of `II_Result`-like classes:
a) limits the possible outcomes to just primitives 
b) looks bad IMO, in JCStress these classes are autogenerated and include terrific names like `ZII_Result` or [`BBBBBBBB_Result`](https://github.com/openjdk/jcstress/blob/ce8a2c7747e0232fac66783c89ea4a58a3819e0f/tests-custom/src/main/java/org/openjdk/jcstress/tests/atomicity/primitives/perbyte/LongAtomicityTest.java#L33)

Reply:
> But this still is a function and it still will do stack pops and stuff?

Yes, but at least it would not interfere with the testing code itself. 

> returning a value is as simple as writing it to a specific register?

Depends on the calling convention, but generally yes.

> Creating a lot of II_Result-like classes

I agree this is an ugly part in JCStress. That is why I proposed to make local state class to be another generic parameter of litmus test class (so that the user can specify the required result class for each test).

Anyway, on the meeting we decided to use one class both for shared fields and result fields.